File: client/src/modules/assetmanager/schemas/captable/security.schemas.ts
/**
 * Security Schemas
 *
 * Zod validation schemas for Security model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/captable_models.py
 * - Schema: /server/apps/assetmanager/schemas/captable_schemas/security_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/captable_subrouters/security_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

/**
 * Security type options - matches backend SecurityType enum
 * Backend: class SecurityType(str, Enum)
 */
export const SecurityTypeEnum = z.enum([
  'common',
  'preferred',
  'convertible',
  'warrant',
  'option',
  'bond',
  'safe',
]);

/**
 * Currency options - matches backend Currency enum
 * Backend: class Currency(str, Enum)
 */
export const CurrencyEnum = z.enum([
  'USD',
  'EUR',
  'GBP',
  'RON',
  'CHF',
  'JPY',
  'CNY',
  'AUD',
  'CAD',
  'SGD',
  'HKD',
]);

/**
 * Anti-dilution protection types - matches backend AntiDilutionType enum
 * Backend: class AntiDilutionType(str, Enum)
 */
export const AntiDilutionTypeEnum = z.enum([
  'none',
  'full_ratchet',
  'weighted_average',
]);

/**
 * Interest rate types - matches backend InterestRateType enum
 * Backend: class InterestRateType(str, Enum)
 */
export const InterestRateTypeEnum = z.enum([
  'fixed',
  'variable',
  'simple',
  'compound',
]);

// ==========================================
// Security Schema (Full Representation)
// ==========================================

/**
 * Security schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class Security(BaseModel)
 */
export const SecuritySchema = z.object({
  id: z.number(),
  funding_round_id: z.number(),
  security_name: z.string().min(1, 'Security name is required').max(255, 'Security name must be less than 255 characters'),
  code: z.string().max(20, 'Code must be less than 20 characters'),
  security_type: SecurityTypeEnum,
  currency: CurrencyEnum,
  issue_price: z.number().nullable(),
  special_terms: z.string().nullable(),

  // Stock Fields
  is_preferred: z.boolean().nullable(),

  // Voting Rights
  has_voting_rights: z.boolean().nullable(),
  voting_ratio: z.number().nullable(),

  // Dividend Rights
  has_dividend_rights: z.boolean().nullable(),
  dividend_rate: z.number().nullable(),
  is_dividend_cumulative: z.boolean().nullable(),

  // Liquidation & Participation
  liquidation_preference: z.number().nullable(),
  has_participation: z.boolean().nullable(),
  participation_cap: z.number().nullable(),
  seniority: z.number().nullable(),
  anti_dilution: AntiDilutionTypeEnum.nullable(),

  // Conversion Rights
  has_conversion_rights: z.boolean().nullable(),
  conversion_ratio: z.number().nullable(),

  // Redemption Rights
  has_redemption_rights: z.boolean().nullable(),
  redemption_term: z.number().nullable(),

  // Convertible Security Fields
  interest_rate: z.number().nullable(),
  interest_rate_type: InterestRateTypeEnum.nullable(),
  interest_period: z.string().nullable(),
  maturity_date: z.string().nullable(), // ISO date string from backend
  valuation_cap: z.number().nullable(),
  conversion_discount: z.number().nullable(),
  conversion_basis: z.string().nullable(),

  // Option-Specific Fields
  option_type: z.string().nullable(),

  // Vesting & Exercise Terms
  vesting_start: z.string().nullable(), // ISO date string from backend
  vesting_months: z.number().nullable(),
  cliff_months: z.number().nullable(),
  vesting_schedule_type: z.string().nullable(),
  exercise_window_days: z.number().nullable(),
  strike_price: z.number().nullable(),
  expiration_date: z.string().nullable(), // ISO date string from backend
  termination_date: z.string().nullable(), // ISO date string from backend

  // Option Pool Management
  pool_name: z.string().nullable(),
  pool_size: z.number().nullable(),
  pool_available: z.number().nullable(),
  is_active: z.boolean().nullable(),

  // Warrant-Specific Fields
  warrant_type: z.string().nullable(),
  is_detachable: z.boolean(),
  deal_context: z.string().nullable(),
  is_transferable: z.boolean(),

  // Shared Option/Warrant Fields
  total_shares: z.number().nullable(),
  issue_rights: z.string().nullable(),
  convert_to: z.string().nullable(),

  // Bond Fields
  principal: z.number().nullable(),
  coupon_rate: z.number().nullable(),
  coupon_frequency: z.string().nullable(),
  principal_frequency: z.string().nullable(),
  tenure_months: z.number().nullable(),
  moratorium_period: z.number().nullable(),

  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new security (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class SecurityCreate(BaseModel)
 */
export const CreateSecuritySchema = z.object({
  // Core Fields (required)
  funding_round_id: z.number(),
  security_name: z.string().min(1, 'Security name is required').max(255, 'Security name must be less than 255 characters'),
  code: z.string().max(20, 'Code must be less than 20 characters'),
  security_type: SecurityTypeEnum,
  currency: CurrencyEnum.default('USD'),
  issue_price: z.number().nullable().optional(),
  special_terms: z.string().nullable().optional(),

  // Stock Fields
  is_preferred: z.boolean().nullable().optional(),

  // Voting Rights
  has_voting_rights: z.boolean().nullable().optional(),
  voting_ratio: z.number().nullable().optional(),

  // Dividend Rights
  has_dividend_rights: z.boolean().nullable().optional(),
  dividend_rate: z.number().nullable().optional(),
  is_dividend_cumulative: z.boolean().nullable().optional(),

  // Liquidation & Participation
  liquidation_preference: z.number().nullable().optional(),
  has_participation: z.boolean().nullable().optional(),
  participation_cap: z.number().nullable().optional(),
  seniority: z.number().nullable().optional(),
  anti_dilution: AntiDilutionTypeEnum.nullable().optional(),

  // Conversion Rights
  has_conversion_rights: z.boolean().nullable().optional(),
  conversion_ratio: z.number().nullable().optional(),

  // Redemption Rights
  has_redemption_rights: z.boolean().nullable().optional(),
  redemption_term: z.number().nullable().optional(),

  // Convertible Security Fields
  interest_rate: z.number().nullable().optional(),
  interest_rate_type: InterestRateTypeEnum.nullable().optional(),
  interest_period: z.string().nullable().optional(),
  maturity_date: z.string().nullable().optional(), // ISO date string
  valuation_cap: z.number().nullable().optional(),
  conversion_discount: z.number().nullable().optional(),
  conversion_basis: z.string().max(50, 'Conversion basis must be less than 50 characters').nullable().optional(),

  // Option-Specific Fields
  option_type: z.string().nullable().optional(),

  // Vesting & Exercise Terms
  vesting_start: z.string().nullable().optional(), // ISO date string
  vesting_months: z.number().nullable().optional(),
  cliff_months: z.number().nullable().optional(),
  vesting_schedule_type: z.string().nullable().optional(),
  exercise_window_days: z.number().nullable().optional(),
  strike_price: z.number().nullable().optional(),
  expiration_date: z.string().nullable().optional(), // ISO date string
  termination_date: z.string().nullable().optional(), // ISO date string

  // Option Pool Management
  pool_name: z.string().nullable().optional(),
  pool_size: z.number().nullable().optional(),
  pool_available: z.number().nullable().optional(),
  is_active: z.boolean().nullable().optional(),

  // Warrant-Specific Fields
  warrant_type: z.string().nullable().optional(),
  is_detachable: z.boolean().default(false),
  deal_context: z.string().nullable().optional(),
  is_transferable: z.boolean().default(false),

  // Shared Option/Warrant Fields
  total_shares: z.number().nullable().optional(),
  issue_rights: z.string().max(50, 'Issue rights must be less than 50 characters').nullable().optional(),
  convert_to: z.string().nullable().optional(),

  // Bond Fields
  principal: z.number().nullable().optional(),
  coupon_rate: z.number().nullable().optional(),
  coupon_frequency: z.string().nullable().optional(),
  principal_frequency: z.string().nullable().optional(),
  tenure_months: z.number().nullable().optional(),
  moratorium_period: z.number().nullable().optional(),
});

/**
 * Schema for updating a security (PUT)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class SecurityUpdate(BaseModel)
 */
export const UpdateSecuritySchema = z.object({
  // Core Fields
  funding_round_id: z.number().optional(),
  security_name: z.string().min(1, 'Security name is required').max(255, 'Security name must be less than 255 characters').optional(),
  code: z.string().max(20, 'Code must be less than 20 characters').optional(),
  security_type: SecurityTypeEnum.optional(),
  currency: CurrencyEnum.optional(),
  issue_price: z.number().nullable().optional(),
  special_terms: z.string().nullable().optional(),

  // Stock Fields
  is_preferred: z.boolean().nullable().optional(),

  // Voting Rights
  has_voting_rights: z.boolean().nullable().optional(),
  voting_ratio: z.number().nullable().optional(),

  // Dividend Rights
  has_dividend_rights: z.boolean().nullable().optional(),
  dividend_rate: z.number().nullable().optional(),
  is_dividend_cumulative: z.boolean().nullable().optional(),

  // Liquidation & Participation
  liquidation_preference: z.number().nullable().optional(),
  has_participation: z.boolean().nullable().optional(),
  participation_cap: z.number().nullable().optional(),
  seniority: z.number().nullable().optional(),
  anti_dilution: AntiDilutionTypeEnum.nullable().optional(),

  // Conversion Rights
  has_conversion_rights: z.boolean().nullable().optional(),
  conversion_ratio: z.number().nullable().optional(),

  // Redemption Rights
  has_redemption_rights: z.boolean().nullable().optional(),
  redemption_term: z.number().nullable().optional(),

  // Convertible Security Fields
  interest_rate: z.number().nullable().optional(),
  interest_rate_type: InterestRateTypeEnum.nullable().optional(),
  interest_period: z.string().nullable().optional(),
  maturity_date: z.string().nullable().optional(),
  valuation_cap: z.number().nullable().optional(),
  conversion_discount: z.number().nullable().optional(),
  conversion_basis: z.string().max(50, 'Conversion basis must be less than 50 characters').nullable().optional(),

  // Option-Specific Fields
  option_type: z.string().nullable().optional(),

  // Vesting & Exercise Terms
  vesting_start: z.string().nullable().optional(),
  vesting_months: z.number().nullable().optional(),
  cliff_months: z.number().nullable().optional(),
  vesting_schedule_type: z.string().nullable().optional(),
  exercise_window_days: z.number().nullable().optional(),
  strike_price: z.number().nullable().optional(),
  expiration_date: z.string().nullable().optional(),
  termination_date: z.string().nullable().optional(),

  // Option Pool Management
  pool_name: z.string().nullable().optional(),
  pool_size: z.number().nullable().optional(),
  pool_available: z.number().nullable().optional(),
  is_active: z.boolean().nullable().optional(),

  // Warrant-Specific Fields
  warrant_type: z.string().nullable().optional(),
  is_detachable: z.boolean().nullable().optional(),
  deal_context: z.string().nullable().optional(),
  is_transferable: z.boolean().nullable().optional(),

  // Shared Option/Warrant Fields
  total_shares: z.number().nullable().optional(),
  issue_rights: z.string().max(50, 'Issue rights must be less than 50 characters').nullable().optional(),
  convert_to: z.string().nullable().optional(),

  // Bond Fields
  principal: z.number().nullable().optional(),
  coupon_rate: z.number().nullable().optional(),
  coupon_frequency: z.string().nullable().optional(),
  principal_frequency: z.string().nullable().optional(),
  tenure_months: z.number().nullable().optional(),
  moratorium_period: z.number().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type SecurityType = z.infer<typeof SecurityTypeEnum>;
export type Currency = z.infer<typeof CurrencyEnum>;
export type AntiDilutionType = z.infer<typeof AntiDilutionTypeEnum>;
export type InterestRateType = z.infer<typeof InterestRateTypeEnum>;
export type Security = z.infer<typeof SecuritySchema>;
export type CreateSecurity = z.infer<typeof CreateSecuritySchema>;
export type UpdateSecurity = z.infer<typeof UpdateSecuritySchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single security
 * Backend equivalent: class SecurityResponse(BaseModel)
 */
export type SecurityResponse = {
  success: boolean;
  data?: Security;
  error?: string;
};

/**
 * Response containing multiple securities
 * Backend equivalent: class SecuritiesResponse(BaseModel)
 */
export type SecuritiesResponse = {
  success: boolean;
  data?: Security[];
  error?: string;
};

// ==========================================
// Type Guards
// ==========================================

/** Human-readable labels for security types */
export const SECURITY_TYPE_LABELS: Record<SecurityType, string> = {
  common: 'Common Shares',
  preferred: 'Preferred Shares',
  convertible: 'Convertible',
  warrant: 'Warrant',
  option: 'Option',
  bond: 'Bond',
  safe: 'SAFE',
};

/**
 * Get human-readable label for a security type
 * @param type Security type enum value
 * @returns Display label string
 */
export const getSecurityTypeLabel = (type: string): string => {
  return SECURITY_TYPE_LABELS[type as SecurityType] || type;
};

/**
 * Check if security is a stock type (common or preferred)
 */
export const isStockSecurity = (security: Security): boolean => {
  return security.security_type === 'common' || security.security_type === 'preferred';
};

/**
 * Check if security is a convertible type
 */
export const isConvertibleSecurity = (security: Security): boolean => {
  return security.security_type === 'convertible';
};

/**
 * Check if security is an option type
 */
export const isOptionSecurity = (security: Security): boolean => {
  return security.security_type === 'option';
};

/**
 * Check if security is a warrant type
 */
export const isWarrantSecurity = (security: Security): boolean => {
  return security.security_type === 'warrant';
};

/**
 * Check if security is a bond type
 */
export const isBondSecurity = (security: Security): boolean => {
  return security.security_type === 'bond';
};

/**
 * Check if security is a SAFE type
 */
export const isSafeSecurity = (security: Security): boolean => {
  return security.security_type === 'safe';
};

File: server/apps/assetmanager/schemas/captable_schemas/security_schemas.py
"""
Security Schemas

Pydantic schemas for the Security model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from datetime import date as date_type
from enum import Enum

# ==========================================
# Enum Types
# ==========================================

class SecurityType(str, Enum):
    """Security type options"""
    COMMON_SHARES = "common"
    PREFERRED_SHARES = "preferred" 
    CONVERTIBLE = "convertible"
    WARRANT = "warrant"
    OPTION = "option"
    BOND = "bond"
    SAFE = "safe"

class Currency(str, Enum):
    """Currency options"""
    USD = "USD"
    EUR = "EUR"
    GBP = "GBP"
    RON = "RON"
    CHF = "CHF"
    JPY = "JPY"
    CNY = "CNY"
    AUD = "AUD"
    CAD = "CAD"
    SGD = "SGD"
    HKD = "HKD"

class AntiDilutionType(str, Enum):
    """Anti-dilution protection types"""
    NONE = "none"
    FULL_RATCHET = "full_ratchet"
    WEIGHTED_AVERAGE = "weighted_average"

class InterestRateType(str, Enum):
    """Interest rate types"""
    FIXED = "fixed"
    VARIABLE = "variable"
    SIMPLE = "simple"
    COMPOUND = "compound"

# ==========================================
# Security Schema (Full Representation)
# ==========================================

class Security(BaseModel):
    """Security schema - full representation"""
    id: int
    funding_round_id: int = Field(description="Associated funding round ID")
    security_name: str = Field(min_length=1, max_length=255, description="Security name")
    code: str = Field(max_length=20, description="Security code")
    security_type: SecurityType = Field(description="Security type")
    currency: Currency = Field(default=Currency.USD, description="Currency")
    issue_price: float | None = Field(None, description="Issue price")
    special_terms: str | None = Field(None, description="Special terms")
    
    # Stock Fields
    is_preferred: bool | None = Field(None, description="Is preferred stock")
    
    # Voting Rights
    has_voting_rights: bool | None = Field(None, description="Has voting rights")
    voting_ratio: float | None = Field(None, description="Voting ratio")
    
    # Dividend Rights
    has_dividend_rights: bool | None = Field(None, description="Has dividend rights")
    dividend_rate: float | None = Field(None, description="Dividend rate")
    is_dividend_cumulative: bool | None = Field(None, description="Is dividend cumulative")
    
    # Liquidation & Participation
    liquidation_preference: float | None = Field(None, description="Liquidation preference")
    has_participation: bool | None = Field(None, description="Has participation rights")
    participation_cap: float | None = Field(None, description="Participation cap")
    seniority: int | None = Field(None, description="Seniority level")
    anti_dilution: AntiDilutionType | None = Field(None, description="Anti-dilution protection")
    
    # Conversion Rights
    has_conversion_rights: bool | None = Field(None, description="Has conversion rights")
    conversion_ratio: float | None = Field(None, description="Conversion ratio")
    
    # Redemption Rights
    has_redemption_rights: bool | None = Field(None, description="Has redemption rights")
    redemption_term: int | None = Field(None, description="Redemption term in months")
    
    # Convertible Security Fields
    interest_rate: float | None = Field(None, description="Interest rate")
    interest_rate_type: InterestRateType | None = Field(None, description="Interest rate type")
    interest_period: str | None = Field(None, description="Interest period")
    maturity_date: date_type | None = Field(None, description="Maturity date")
    valuation_cap: float | None = Field(None, description="Valuation cap")
    conversion_discount: float | None = Field(None, description="Conversion discount")
    conversion_basis: str | None = Field(None, max_length=50, description="Conversion basis")
    
    # Option-Specific Fields
    option_type: str | None = Field(None, description="Option type (esop, vsop, sar)")
    
    # Vesting & Exercise Terms
    vesting_start: date_type | None = Field(None, description="Vesting start date")
    vesting_months: int | None = Field(None, description="Vesting period in months")
    cliff_months: int | None = Field(None, description="Cliff period in months")
    vesting_schedule_type: str | None = Field(None, description="Vesting schedule type")
    exercise_window_days: int | None = Field(None, description="Exercise window in days")
    strike_price: float | None = Field(None, description="Strike price")
    expiration_date: date_type | None = Field(None, description="Expiration date")
    termination_date: date_type | None = Field(None, description="Termination date")
    
    # Option Pool Management
    pool_name: str | None = Field(None, description="Pool name")
    pool_size: float | None = Field(None, description="Pool size")
    pool_available: float | None = Field(None, description="Pool available amount")
    is_active: bool | None = Field(None, description="Is active")
    
    # Warrant-Specific Fields
    warrant_type: str | None = Field(None, description="Warrant type")
    is_detachable: bool = Field(default=False, description="Is detachable")
    deal_context: str | None = Field(None, description="Deal context")
    is_transferable: bool = Field(default=False, description="Is transferable")
    
    # Shared Option/Warrant Fields
    total_shares: float | None = Field(None, description="Total shares")
    issue_rights: str | None = Field(None, max_length=50, description="Issue rights")
    convert_to: str | None = Field(None, description="Convert to")
    
    # Bond Fields
    principal: float | None = Field(None, description="Principal amount")
    coupon_rate: float | None = Field(None, description="Coupon rate")
    coupon_frequency: str | None = Field(None, description="Coupon frequency")
    principal_frequency: str | None = Field(None, description="Principal frequency")
    tenure_months: int | None = Field(None, description="Tenure in months")
    moratorium_period: int | None = Field(None, description="Moratorium period")
    
    created_at: datetime
    updated_at: datetime | None = None
    
    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class SecurityCreate(BaseModel):
    """Schema for creating a new security"""
    funding_round_id: int = Field(description="Associated funding round ID")
    security_name: str = Field(min_length=1, max_length=255, description="Security name")
    code: str = Field(max_length=20, description="Security code")
    security_type: SecurityType = Field(description="Security type")
    currency: Currency = Field(default=Currency.USD, description="Currency")
    issue_price: float | None = Field(None, description="Issue price")
    special_terms: str | None = Field(None, description="Special terms")
    
    # Stock Fields
    is_preferred: bool | None = None
    
    # Voting Rights
    has_voting_rights: bool | None = None
    voting_ratio: float | None = None
    
    # Dividend Rights
    has_dividend_rights: bool | None = None
    dividend_rate: float | None = None
    is_dividend_cumulative: bool | None = None
    
    # Liquidation & Participation
    liquidation_preference: float | None = None
    has_participation: bool | None = None
    participation_cap: float | None = None
    seniority: int | None = None
    anti_dilution: AntiDilutionType | None = None

    # Conversion Rights
    has_conversion_rights: bool | None = None
    conversion_ratio: float | None = None

    # Redemption Rights
    has_redemption_rights: bool | None = None
    redemption_term: int | None = None

    # Convertible Security Fields
    interest_rate: float | None = None
    interest_rate_type: InterestRateType | None = None
    interest_period: str | None = None
    maturity_date: date_type | None = None
    valuation_cap: float | None = None
    conversion_discount: float | None = None
    conversion_basis: str | None = Field(None, max_length=50)
    
    # Option-Specific Fields
    option_type: str | None = None
    
    # Vesting & Exercise Terms
    vesting_start: date_type | None = None
    vesting_months: int | None = None
    cliff_months: int | None = None
    vesting_schedule_type: str | None = None
    exercise_window_days: int | None = None
    strike_price: float | None = None
    expiration_date: date_type | None = None
    termination_date: date_type | None = None
    
    # Option Pool Management
    pool_name: str | None = None
    pool_size: float | None = None
    pool_available: float | None = None
    is_active: bool | None = None
    
    # Warrant-Specific Fields
    warrant_type: str | None = None
    is_detachable: bool = Field(default=False)
    deal_context: str | None = None
    is_transferable: bool = Field(default=False)
    
    # Shared Option/Warrant Fields
    total_shares: float | None = None
    issue_rights: str | None = Field(None, max_length=50)
    convert_to: str | None = None
    
    # Bond Fields
    principal: float | None = None
    coupon_rate: float | None = None
    coupon_frequency: str | None = None
    principal_frequency: str | None = None
    tenure_months: int | None = None
    moratorium_period: int | None = None

class SecurityUpdate(BaseModel):
    """Schema for updating a security"""
    funding_round_id: int | None = None
    security_name: str | None = Field(None, min_length=1, max_length=255)
    code: str | None = Field(None, max_length=20)
    security_type: SecurityType | None = None
    currency: Currency | None = None
    issue_price: float | None = None
    special_terms: str | None = None
    
    # Stock Fields
    is_preferred: bool | None = None
    
    # Voting Rights
    has_voting_rights: bool | None = None
    voting_ratio: float | None = None
    
    # Dividend Rights
    has_dividend_rights: bool | None = None
    dividend_rate: float | None = None
    is_dividend_cumulative: bool | None = None
    
    # Liquidation & Participation
    liquidation_preference: float | None = None
    has_participation: bool | None = None
    participation_cap: float | None = None
    seniority: int | None = None
    anti_dilution: AntiDilutionType | None = None

    # Conversion Rights
    has_conversion_rights: bool | None = None
    conversion_ratio: float | None = None

    # Redemption Rights
    has_redemption_rights: bool | None = None
    redemption_term: int | None = None

    # Convertible Security Fields
    interest_rate: float | None = None
    interest_rate_type: InterestRateType | None = None
    interest_period: str | None = None
    maturity_date: date_type | None = None
    valuation_cap: float | None = None
    conversion_discount: float | None = None
    conversion_basis: str | None = Field(None, max_length=50)
    
    # Option-Specific Fields
    option_type: str | None = None
    
    # Vesting & Exercise Terms
    vesting_start: date_type | None = None
    vesting_months: int | None = None
    cliff_months: int | None = None
    vesting_schedule_type: str | None = None
    exercise_window_days: int | None = None
    strike_price: float | None = None
    expiration_date: date_type | None = None
    termination_date: date_type | None = None
    
    # Option Pool Management
    pool_name: str | None = None
    pool_size: float | None = None
    pool_available: float | None = None
    is_active: bool | None = None
    
    # Warrant-Specific Fields
    warrant_type: str | None = None
    is_detachable: bool | None = None
    deal_context: str | None = None
    is_transferable: bool | None = None
    
    # Shared Option/Warrant Fields
    total_shares: float | None = None
    issue_rights: str | None = Field(None, max_length=50)
    convert_to: str | None = None
    
    # Bond Fields
    principal: float | None = None
    coupon_rate: float | None = None
    coupon_frequency: str | None = None
    principal_frequency: str | None = None
    tenure_months: int | None = None
    moratorium_period: int | None = None

# ==========================================
# Response Types
# ==========================================

class SecurityResponse(BaseModel):
    """Response containing a single security"""
    success: bool
    data: Security | None = None
    error: str | None = None

class SecuritiesResponse(BaseModel):
    """Response containing multiple securities"""
    success: bool
    data: list[Security] | None = None
    error: str | None = None